Add documentation on feature flag toggle functionality (#5843)
alphagov/govuk-frontend: docs/contributing/rebrand-feature-flag.md

# The 'rebrand' feature flag

While we work on the rebrand of GOV.UK Frontend, you can use the feature flag toggling system to help with sharing and visualising changes and testing rebranded components.

## How to use the UI

> You can see the feature flag toggling UI at the top of most pages in the review app as a details element. It does not appear on component previews. When you open it, you'll see a form with a checkbox labelled 'rebrand feature flag on'. By checking this box and clicking 'submit', you'll reload the page. If you're looking at a page on the review app with components impacted by the rebrand eg: the header, you'll see the rebranded version of that component. Unchecking it and resubmitting the form will turn the feature flag off.

Author: I've added a section called 'Things that change when the rebrand flag is on' further up the page to try to address this. Let me know what you think!

Additionally, when on a component landing page eg: '/components/header', you can click the link just under the feature flag form labelled 'show all flag states'. Clicking this will show two example iframes for each component example, one with the new brand and one with the old brand. Clicking the link again or navigating to a new page will turn this feature off.

## How it works

### Feature flags middleware

This feature is controlled by the app's feature flag middleware which does two things:

1. Upon making a post request to `/set-rebrand`, which the feature flag form does, the middleware sets a `use_rebrand` cookie and redirects back to the page the user was just on
2. Sets 3 local variables:

- `useRebrand` which is set based on the presence of either a `rebrandOverride` query parameter or the state of the `use_rebrand` cookie if the override param isn't present
- `ShowAllFlagStates` which is set based on the presence of a `showAllFlagStates` query parameter. The link under the feature flag form sets this parameter
- `exampleStates` which is either `[true, false]` if `showAllFlagStates` is `true`, else an array containing `useRebrand` as one item

> These local variables are passed to the app views where they're used to display or not display rebranded states of things. `useRebrand` is also used by the main express app to automatically set the context of component examples so that if a given component has a `rebrand` param, its state is automatically set.

Author: I've added a section called 'Things that change when the rebrand flag is on' further up the page to try to address this. Let me know what you think!

# The 'rebrand' feature flag

> [!NOTE]

> > This documentation is for the GOV.UK Design System team and is not relevant to users wanting to contribute to GOV.UK Frontend or the GOV.UK Design System

Reviewer: I wonder if we need to separate out things like release docs and this which are for the team, and, like, browser support and application architecture, which are relevant to folks trying to do contributions. Not necessarily something for this commit though.

Author: It's tricky because a lot of the stuff that applies to us in 'contributing' also applies to anyone trying to do contribution. In theory we're an open source project so anyone can contribute to us in any form, so it's all relevant info for using the codebase.

I think I'm actually gonna remove this. It feels a bit gratuitous looking at it again. I think I put it here because this is very opinionated documentation but that's not a reason to make a big song and dance.

> Whilst we work on the rebrand of GOV.UK Frontend, you can use the feature flag toggling system to help with sharing and visualising changes and testing rebranded components.

Reviewer: ```suggestion
While we work on the rebrand of GOV.UK Frontend, you can use the feature flag toggling system to help with sharing and visualising changes and testing rebranded components.
```

As per style guide

## How to use the UI

> You can see the feature flag toggling UI at the top of all pages in the review app as a details element, except component previews. When you open it, you'll see a form with a checkbox labelled 'rebrand feature flag on'. By checking this box and clicking 'submit', you'll reload the page. If you're looking at a page on the review app with components impacted by the rebrand eg: the header, you'll see the rebranded version of that component. Unchecking it and resubmitting the form will turn the feature flag off.

Reviewer: ```suggestion
You can see the feature flag toggling UI at the top of most pages in the review app as a details element. It does not appear on component previews. When you open it, you'll see a form with a checkbox labelled 'rebrand feature flag on'. By checking this box and clicking 'submit', you'll reload the page. If you're looking at a page on the review app with components impacted by the rebrand eg: the header, you'll see the rebranded version of that component. Unchecking it and resubmitting the form will turn the feature flag off.
```

> Additionally, when on a component landing page eg: '/components/header', you can click the link just under the feature flag form labelled 'show all flag states'. Clicking this will show two example iframes for each component example, one with the new brand and one with the old brand. Clicking the link again or navigating to a new page will turn this feature off.

Reviewer: Oooh! That's nice!
